remove prototype mode

akphi opened the pull request:
# Summary

- [x] Multiple mode-specific area in Gitlab configuration is no longer supported, instead, we move the `appConfig` and `serverConfig` to top level in `Gitlab configuration`
- [x] Rename `ProjectAuthorizationAction` -> `ProjectAuthorizedAction`
- [x] Elimite `GitLabMode`, `GitLabModeInfo`, `GitLabTokenManager`
- [x] Add field `projectIdPrefix` to `GitlabConfiguration` to configure `project ID prefix`, e.g. `PROD-1234`, `PROTOTYPE-1235`, `PREFIX-123`, if no prefix is supplied, the project ID will be similar to Gitlab ID, e.g. `1235`
- [x] Ignore `modes` as inputs to APIs, these will be deprecated and hidden: **NOTE**: our current version of `Swagger` does not allow deprecating the parameter, we will just hide and ignore it.
- [x] Deprecate `ProjectType`: remove logic involving `ProjectType`:
  - [x] Checking for `non-production` when updating dependencies in `ProjectStructure`
  - [x] Checking for prototype projects when fetching versions and default to return empty list
- [x] Add server features configuration `LegendSDLCServerFeaturesConfiguration` to manage supported actions by the server to take care of behavior determined by `ProjectType` previously
  - [x] Create `ServerResource` to deprecate `InfoResource` and create endpoint `/server/features` to get the list of supported features
  - [x] Support actions including: project creation, version creation 


# TODO

- [ ] Remove `ProjectType` enum and change all usages of `ProjectType` in models to `String`. Could we completely remove `projectType` in these places, how would that affect (de)serialization?
- [ ] Remove methods using `ProjectType` in
- [ ] Create new project structure extensions in existing infrastructure and alter `gitlab-ci` files there accordingly to allow/disallow publish, etc.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/SimpleProjectConfiguration.java:
```diff
@@ -86,14 +85,10 @@ public void setProjectId(String projectId)
     }
 
     @Override
+    @JsonIgnore
```
I'm not sure if you're okay with this @kevin-m-knight-gs

kevin-m-knight-gs replied:
What is it you're trying to do with the annotation?

akphi replied:
Addressed below

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/config/LegendSDLCServerFeaturesConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+// Copyright 2020 Goldman Sachs
```
```suggestion
// Copyright 2022 Goldman Sachs
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenHelper.java:
```diff
@@ -0,0 +1,66 @@
+// Copyright 2020 Goldman Sachs
```
```suggestion
// Copyright 2022 Goldman Sachs
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/resources/ServerResource.java:
```diff
@@ -0,0 +1,58 @@
+// Copyright 2020 Goldman Sachs
```
```suggestion
// Copyright 2022 Goldman Sachs
```

kevin-m-knight-gs reviewed (changes requested):
Make sure to use 2022 in the copyright headers of new files.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabProjectId.java:
```diff
@@ -51,30 +45,30 @@ public boolean equals(Object other)
-        GitLabProjectId that = (GitLabProjectId)other;
-        return (this.mode == that.mode) && (this.gitLabId == that.gitLabId);
+        GitLabProjectId that = (GitLabProjectId) other;
+        return ((this.prefix == null && that.prefix == null) || this.prefix != null && this.prefix.equals(that.prefix)) && (this.gitLabId == that.gitLabId);
```
```suggestion
        return Objects.equals(this.prefix, that.prefix) && (this.gitLabId == that.gitLabId);
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabProjectId.java:
```diff
@@ -51,30 +45,30 @@ public boolean equals(Object other)
     public int hashCode()
     {
-        return this.mode.hashCode() ^ this.gitLabId;
+        return (this.prefix != null ? this.prefix.hashCode() : 0) ^ this.gitLabId;
```
```suggestion
        return Objects.hashCode(this.prefix) ^ this.gitLabId;
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabProjectId.java:
```diff
@@ -123,12 +90,12 @@ private static int parseGitLabId(String projectId, int separatorIndex)
         catch (NumberFormatException e)
         {
-            throw new IllegalArgumentException("Invalid project id: " + projectId);
+            throw new IllegalArgumentException("Invalid project ID: " + projectId);
```
```suggestion
            throw new IllegalArgumentException("Invalid project id: " + projectId);
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabServerHealthCheck.java:
```diff
@@ -15,26 +15,23 @@
     {
         // TODO do a more substantive health check
-        return this.gitLabModeInfos.isEmpty() ? Result.unhealthy("No GitLab modes available") : Result.healthy();
+        return Result.healthy();
```
At the very least, let's maintain the current level of validation by checking that server and app info have been configured. If possible, we should check that the GitLab server is accessible.

akphi replied:
I will only check for the app info for now, will leave a comment regarding your suggestion

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java:
```diff
@@ -585,63 +537,50 @@ protected LegendSDLCServerException buildException(Exception e, Supplier<String>
-                (defaultMessage == null) ?
-                        null :
-                        ex -> Optional.ofNullable(defaultMessage.get()).map(m -> StringTools.appendThrowableMessageIfPresent(m, ex)).orElse(null));
+        return buildException(e, (forbiddenMessage == null) ? null : ex -> Optional.ofNullable(forbiddenMessage.get()).map(m -> StringTools.appendThrowableMessageIfPresent(m, ex)).orElse(null), (notFoundMessage == null) ? null : ex -> notFoundMessage.get(), (defaultMessage == null) ? null : ex -> Optional.ofNullable(defaultMessage.get()).map(m -> StringTools.appendThrowableMessageIfPresent(m, ex) ...
```
This was easier to read in its previous multi-line format.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java:
```diff
@@ -585,63 +537,50 @@ protected LegendSDLCServerException buildException(Exception e, Supplier<String>
-        return processException(e,
-                Function.identity(),
-                glae ->
+        return processException(e, Function.identity(), glae ->
```
I think the code is more readable in this case when there's one argument per line.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/BaseGitLabApi.java:
```diff
@@ -826,10 +765,7 @@ protected MergeRequest getReviewMergeRequest(MergeRequestApi mergeRequestApi, Gi
-                    () -> "User " + getCurrentUser() + " is not allowed to access review " + reviewId + " in project " + projectId,
-                    () -> "Unknown review in project " + projectId + ": " + reviewId,
-                    () -> "Error accessing review " + reviewId + " in project " + projectId);
+            throw buildException(e, () -> "User " + getCurrentUser() + " is not allowed to access review " + reviewId + " in project " + projectId, () -> "Unknown review in project " + projectId + ": " + reviewId, () -> "Error accessing review " + reviewId + " in project " + projectId);
```
This was easier to read in its previous multi-line format.

kevin-m-knight-gs reviewed (changes requested):
There are a number of places, especially in exception handling, where you've compressed multi-line code into a single line making it much harder to read. I put comments on some of the places, but there are lots more that I didn't specifically comment on. Can you return them all to their previous multi-line format?

akphi commented:
@kevin-m-knight-gs I use IntelliJ auto-formatter. I will try undo as much as I could.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabConfiguration.java:
```diff
@@ -83,14 +93,14 @@ public Visibility getNewProjectVisibility()
 
     @JsonCreator
-    public static GitLabConfiguration newGitLabConfiguration(@JsonProperty("projectTag") String projectTag, @JsonProperty("auth") AuthConfiguration authConfig, @JsonProperty("uat") ModeConfiguration uatConfig, @JsonProperty("prod") ModeConfiguration prodConfig, @JsonProperty("newProjectVisibility") NewProjectVisibility newProjectVisibility, @JsonProperty("gitlabAuthorizers") List<GitLabAuthorizer ...
+    public static GitLabConfiguration newGitLabConfiguration(@JsonProperty("projectTag") String projectTag, @JsonProperty("projectIdPrefix") String projectIdPrefix, @JsonProperty("auth") AuthConfiguration authConfig, @JsonProperty("server") ServerConfiguration serverConfig, @JsonProperty("app") AppConfiguration appConfig, @JsonProperty("newProjectVisibility") NewProjectVisibility newProjectVisibi ...
```
Done!

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenHelper.java:
```diff
@@ -0,0 +1,66 @@
+{
+    public static Token.TokenBuilder encodeGitLabToken(Token.TokenBuilder builder, GitLabToken token, GitLabAppInfo appInfo)
+    {
+        if (token != null)
```
See the below comment about maintaining backward compatibility. Perhaps this should start with an integer indicating how many (1 or 0) tokens are being encoded.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java:
```diff
@@ -55,76 +58,66 @@ public boolean equals(Object other)
+    public boolean gitLabOAuthCallback(String code)
     {
-        return this.tokenManager.gitLabOAuthCallback(mode, code);
+        GitLabToken token = GitLabToken.newGitLabToken(Constants.TokenType.OAUTH2_ACCESS, GitLabOAuthAuthenticator.newAuthenticator(appInfo).getOAuthTokenFromAuthCode(code));
+        return !token.equals(this.token);
```
The token needs to be set on the field.

akphi replied:
My bad, I messed up my code reading 😢 😢 😢

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java:
```diff
@@ -55,76 +58,66 @@ public boolean equals(Object other)
     }
 
-    private static GitLabTokenManager possiblyInitializeTokenManager(GitLabTokenManager tokenManager, OidcProfile profile)
+    private void possiblyInitializeGitLabToken(GitLabAppInfo appInfo, OidcProfile profile)
```
```suggestion
    private static GitLabToken possiblyInitializeGitLabToken(GitLabAppInfo appInfo, OidcProfile profile)
```

akphi replied:
Don't think I will do this. I know it looks like we diverge from the original but since this method possibly return `null` (in case you want it to return `GitLabToken`), and we need to set its return value in the constructor, so something like 

```java
construtor(...)
{
        super(profile, userId, creationTime);
        this.appInfo = appInfo;
        this.token = token;
        GitLabToken newToken = this.possiblyInitializeGitLabToken(appInfo, profile);
         if (newToken != null)
         {
             this.token = newToken;
         }
    }
```

I think that's clunky, so there's no real benefit to say this method must return a value.

kevin-m-knight-gs replied:
I think you're misunderstanding the desired logic here. It should simply be:

`this.token = possiblyInitializeGitLabToken(appInfo, profile);`

The main point of possiblyInitializeGitLabToken is to check whether the token from the profile (i.e., the one from authentication) can be used in the session.  There's no guarantee that it is suitable for connecting to the GitLab API. That's what we're ensuring here. If it returns null, then we want `this.token` to be null.

Another thing is that when possible it's best to use static methods to initialize values in a constructor. It ensures there's no conflict with other state on the instance. It also makes it clear in the constructor exactly what information is being used in the initialization.

akphi replied:
Gotcha! Done now

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java:
```diff
@@ -137,20 +130,14 @@ private static GitLabTokenManager possiblyInitializeTokenManager(GitLabTokenMana
-                                });
+                        if (issuer.equals(appInfo.getServerInfo().getGitLabURLString()))
+                        {
+                            this.setGitLabToken(GitLabToken.newGitLabToken(TokenType.OAUTH2_ACCESS, accessToken.getValue()));
+                            LOGGER.debug("Storing access token from OpenID Connect (OIDC) profile");
```
```suggestion
                            LOGGER.debug("Storing access token from OpenID Connect (OIDC) profile");
                            return GitLabToken.newGitLabToken(TokenType.OAUTH2_ACCESS, accessToken.getValue());
```

akphi replied:
Please see my response above

kevin-m-knight-gs reviewed (changes requested):
Lots of comments here. Many of them relate to formatting, but there are a many substantive comments as well.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/GitLabConfiguration.java:
```diff
@@ -83,14 +94,42 @@ public Visibility getNewProjectVisibility()
+        @JsonProperty("gitlabAuthorizers") List<GitLabAuthorizer> gitLabAuthorizers)
     {
-        return new GitLabConfiguration(projectTag, authConfig, uatConfig, prodConfig, newProjectVisibility, gitLabAuthorizers);
+        ServerConfiguration _serverConfig = serverConfig;
+        AppConfiguration _appConfig = appConfig;
+        String _projectIdPrefix = projectIdPrefix;
+        if (uatConfig != null && prodConfig != null)
+        {
+            throw new IllegalArgumentException("Configuration with multiple Gitlab modes is no longer supported");
+        }
+        else if (uatConfig != null)
+        {
+            _serverConfig = uatConfig.getServerConfiguration();
+            _appConfig = uatConfig.getAppConfiguration();
+            _projectIdPrefix = "UAT";
+        }
+        else if (prodConfig != null)
+        {
+            _serverConfig = prodConfig.getServerConfiguration();
+            _appConfig = prodConfig.getAppConfiguration();
+            _projectIdPrefix = "PROD";
+        }
+        return new GitLabConfiguration(projectTag, _projectIdPrefix, authConfig, _serverConfig, _appConfig, newProjectVisibility, gitLabAuthorizers);
```
Done!

akphi commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/gitlab/api/GitLabProjectConfigurationApiTestResource.java:
```diff
@@ -76,8 +73,6 @@ public void runUserAndGroupWorkspaceProjectConfigurationTest()
         ProjectConfiguration projectConfiguration = gitLabProjectConfigurationApi.getUserWorkspaceProjectConfiguration(projectId, workspaceOneId);
 
         Assert.assertNotNull(projectConfiguration);
-        Assert.assertTrue(projectConfiguration.getProjectId().startsWith("PROD-"));
```
Maybe, this requires us to configure the `projectIdPrefix`, which is configured outside the context of this class, but the assertion is in this method, this will require me to update both `TestGitLabServerProjectConfigurationApis` and `IntegrationTestGitLabProjectConfigurationApis`, I'm not sure if we should do that or we could take that the unit test already taken care of this in a way?

akphi commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/gitlab/TestGitLabProjectId.java:
```diff
@@ -26,55 +25,49 @@
     public void testGetProjectIdString()
     {
-        for (GitLabMode mode : GitLabMode.values())
+        String mode = "SOMEPREFIX";
```
of course! Done!

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/resources/GitLabAuthResource.java:
```diff
@@ -62,53 +61,35 @@ public Object callback(@QueryParam("code") String code, @QueryParam("state") Str
+            }
+            catch (GitLabAuthAccessException e)
+            {
+                // tolerate access exceptions
+                getLogger().error("Access exception occurred while checking authorization", e);
+                return true;
```
done!

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabAuthorizer.java:
```diff
@@ -15,9 +15,9 @@
 public interface GitLabAuthorizer
 {
-    GitLabToken authorize(Session session, GitLabModeInfo modeInfo);
+    GitLabToken authorize(Session session, GitLabAppInfo appInfo);
```
I don't really want to keep `GitLabModeInfo` around just because of this. Plus I think `appInfo` makes sense and unless we do something like the following:

```java

default GitLabToken authorize(Session session, GitLabAppInfo appInfo)
{
    return this.authorize(Session session, <create a mock modeInfo object here>);
}
```

Which is a bit less than ideal ...

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabUserContext.java:
```diff
@@ -28,151 +27,120 @@
     {
-        GitLabApi api = this.apiCache.get(mode);
-        if (api == null)
+        GitLabApi gitLabApi = this.api;
```
yep! fixed now

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabPersonalAccessTokenSession.java:
```diff
@@ -52,94 +55,75 @@ public boolean equals(Object other)
+    public boolean gitLabOAuthCallback(String code)
     {
-        return this.tokenManager.gitLabOAuthCallback(mode, code);
+        GitLabToken token = GitLabToken.newGitLabToken(Constants.TokenType.OAUTH2_ACCESS, GitLabOAuthAuthenticator.newAuthenticator(appInfo).getOAuthTokenFromAuthCode(code));
+        return !token.equals(this.token);
```
My bad, I messed up my code reading 😢

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabKerberosSession.java:
```diff
@@ -49,70 +51,60 @@ public boolean equals(Object other)
+    public boolean gitLabOAuthCallback(String code)
     {
-        return this.tokenManager.gitLabOAuthCallback(mode, code);
+        GitLabToken token = GitLabToken.newGitLabToken(Constants.TokenType.OAUTH2_ACCESS, GitLabOAuthAuthenticator.newAuthenticator(appInfo).getOAuthTokenFromAuthCode(code));
+        return !token.equals(this.token);
```
My bad, I messed up my code reading 😢 😢

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenHelper.java:
```diff
@@ -0,0 +1,66 @@
+import org.finos.legend.sdlc.server.gitlab.GitLabAppInfo;
+import org.gitlab4j.api.Constants;
+
+public class GitLabTokenHelper
```
Yes, done!

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenHelper.java:
```diff
@@ -0,0 +1,66 @@
+
+    public static GitLabToken decodeGitLabToken(Token.TokenReader reader, GitLabAppInfo appInfo)
+    {
+        if (!reader.hasRemaining())
+        {
+            return null;
+        }
+
```
yep, I also has the feeling that I did something fishy there. I have done it the way you mentioned in the comment for `encode`, it will add an integer `0` when there is no token and `1` when there is, and will throw when the count is out of that range.

akphi commented:
@kevin-m-knight-gs I have addressed all the formatting shift I created. Really sorry about that. To lessen the noise during review I think you can turn on `ignore whitespace changes` mode in Github, nevertheless, I wonder if we could set up an auto-formatter that will do all formatting for us automatically so there would be no dispute over this kind of thing in future - I do love to hit `Auto Format` in IntelliJ... so I really don't mind configure IntelliJ using some config and trust it.

I tried to install [Checkstyle-IDEA plugin](https://plugins.jetbrains.com/plugin/1065-checkstyle-idea) and point it at our `checkstyle.xml` file but auto-format function still squashing all the multi-line parameters blob that we have.

Last but not least, I have addressed most of the more substantial comments, except for a few where I have stated the reason why I don't follow them. We can discuss those later!

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenManager.java:
```diff
@@ -50,169 +46,107 @@ public boolean equals(Object other)
-                    builder.putString(token.getTokenType().toString());
-                    builder.putString(token.getToken());
-                }
-            }
-            return builder;
+            return;
+        }
+        if (tokenCount > 1 || tokenCount < 0)
+        {
+            throw new IllegalStateException("Unexpected value (" + tokenCount + ") for token count");
```
I would say that you should loop through all the tokens, regardless of how many there are. If there are extra tokens in the reader that don't match the appId, they can simply be discarded. They aren't harmful. (It's important, though, that you read through all tokens from the reader, so that it's in an appropriate state for whatever reads from it next.)

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabPersonalAccessTokenSession.java:
```diff
@@ -66,78 +65,52 @@ public int hashCode()
+    public GitLabToken getGitLabToken()
     {
-        return this.tokenManager.isValidMode(mode);
+        return tokenManager.getGitLabToken();
```
```suggestion
        return this.tokenManager.getGitLabToken();
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabPersonalAccessTokenSession.java:
```diff
@@ -66,78 +65,52 @@ public int hashCode()
     }
 
-    private static GitLabTokenManager possiblyInitializeTokenManager(GitLabTokenManager tokenManager, GitlabPersonalAccessTokenProfile profile)
+    private static GitLabTokenManager possiblyInitializeGitLabToken(GitLabTokenManager tokenManager, GitlabPersonalAccessTokenProfile profile)
```
We should probably revert the name to possiblyInitializeTokenManager.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabPersonalAccessTokenSession.java:
```diff
@@ -66,78 +65,52 @@ public int hashCode()
             String token = profile.getPersonalAccessToken();
 
-            if (token != null && profile.getGitlabHost() != null)
+            if (tokenManager.getAppInfo().getServerInfo().getHost().equals(profile.getGitlabHost()))
```
You should probably still keep the token != null check.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java:
```diff
@@ -69,60 +68,42 @@ public int hashCode()
+    public GitLabToken getGitLabToken()
     {
-        return this.tokenManager.isValidMode(mode);
+        return tokenManager.getGitLabToken();
```
```suggestion
        return this.tokenManager.getGitLabToken();
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenManager.java:
```diff
@@ -50,169 +46,107 @@ public boolean equals(Object other)
     }
 
-    StringBuilder appendTokenInfo(StringBuilder builder)
+    Token.TokenBuilder encodeGitLabToken(Token.TokenBuilder builder)
```
Let's revert the name to simply "encode".

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabOidcSession.java:
```diff
@@ -69,60 +68,42 @@ public int hashCode()
     }
 
-    private static GitLabTokenManager possiblyInitializeTokenManager(GitLabTokenManager tokenManager, OidcProfile profile)
+    private static GitLabTokenManager possiblyInitializeGitLabToken(GitLabTokenManager tokenManager, OidcProfile profile)
```
Let's revert the name to "possiblyInitializeTokenManager".

kevin-m-knight-gs reviewed (changes requested):
More comments, particularly around token management.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/auth/GitLabTokenManager.java:
```diff
@@ -50,169 +46,107 @@ public boolean equals(Object other)
-            GitLabToken oldToken = this.tokens.put(mode, token);
-            return !token.equals(oldToken);
-        }
+        return builder.append("token=[").append(this.token != null ? this.token.toString() : "").append("]");
```
So when `token == null` do you mean to put `token=` or `token=null`?

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitlabWorkflowApi.java:
```diff
@@ -178,22 +179,18 @@ public Workflow getWorkflow(String workflowId)
-                        // resolve revision alias when possible
-                        .map(revisionId -> resolveRevisionId(revisionId, getRevisionAccessContext()))
-                        .collect(Collectors.toSet());
+                Set<String> revisionIdSet = (revisionIds == null) ? Collections.emptySet() : StreamSupport.stream(revisionIds.spliterator(), false)
+                    // make sure various synonymous aliases are grouped together properly
+                    .filter(Objects::nonNull).map(revisionId ->
+                    {
+                        RevisionAlias alias = getRevisionAlias(revisionId);
+                        return (alias == RevisionAlias.REVISION_ID) ? revisionId : alias.getValue();
+                    }).distinct()
+                    // resolve revision alias when possible
+                    .map(revisionId -> resolveRevisionId(revisionId, getRevisionAccessContext())).collect(Collectors.toSet());
```
This is a weird formatting mix. Of course, so was the previous formatting. Some stream operations get their own line, others are grouped on a single line. For long chains of stream operations, I think it's clearest to have each operation on a separate line.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitlabWorkflowApi.java:
```diff
@@ -210,31 +207,28 @@ public Workflow getWorkflow(String workflowId)
-                                try
-                                {
-                                    return withRetries(() -> pipelineApi.getPipeline(this.projectId.getGitLabId(), p.getId()));
-                                }
-                                catch (Exception ignore)
-                                {
-                                    // ignore exception
-                                }
-                            }
-                            return p;
-                        })
-                        .map(p -> fromGitLabPipeline(this.projectId.toString(), p))
-                        .collect(PagerTools.listCollector(pager, limit));
+                            // ignore exception
+                        }
+                    }
+                    return p;
+                }).map(p -> fromGitLabPipeline(this.projectId.toString(), p)).collect(PagerTools.listCollector(pager, limit));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabBuildApi.java:
```diff
@@ -178,22 +179,18 @@ public Build getBuild(String buildId)
-                        // resolve revision alias when possible
-                        .map(revisionId -> resolveRevisionId(revisionId, getRevisionAccessContext()))
-                        .collect(Collectors.toSet());
+                Set<String> revisionIdSet = (revisionIds == null) ? Collections.emptySet() : StreamSupport.stream(revisionIds.spliterator(), false)
+                    // make sure various synonymous aliases are grouped together properly
+                    .filter(Objects::nonNull).map(revisionId ->
+                    {
+                        RevisionAlias alias = getRevisionAlias(revisionId);
+                        return (alias == RevisionAlias.REVISION_ID) ? revisionId : alias.getValue();
+                    }).distinct()
+                    // resolve revision alias when possible
+                    .map(revisionId -> resolveRevisionId(revisionId, getRevisionAccessContext())).collect(Collectors.toSet());
```
Since you're reformatting, can you put each stream operation on its own line?

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabBuildApi.java:
```diff
@@ -210,31 +207,28 @@ public Build getBuild(String buildId)
-                                try
-                                {
-                                    return withRetries(() -> pipelineApi.getPipeline(this.projectId.getGitLabId(), p.getId()));
-                                }
-                                catch (Exception ignore)
-                                {
-                                    // ignore exception
-                                }
-                            }
-                            return p;
-                        })
-                        .map(p -> fromGitLabPipeline(this.projectId, p))
-                        .collect(PagerTools.listCollector(pager, limit));
+                            // ignore exception
+                        }
+                    }
+                    return p;
+                }).map(p -> fromGitLabPipeline(this.projectId, p)).collect(PagerTools.listCollector(pager, limit));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabEntityApi.java:
```diff
@@ -734,9 +729,7 @@ private ProjectFileOperation entityChangeToFileOperation(EntityChange change, Pr
-        return accessContext.getFilesInDirectory(sourceDirectory.getDirectory())
-                .filter(f -> sourceDirectory.isPossiblyEntityFilePath(f.getPath()))
-                .map(f -> new EntityProjectFile(sourceDirectory, f));
+        return accessContext.getFilesInDirectory(sourceDirectory.getDirectory()).filter(f -> sourceDirectory.isPossiblyEntityFilePath(f.getPath())).map(f -> new EntityProjectFile(sourceDirectory, f));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabRevisionApi.java:
```diff
@@ -230,10 +228,7 @@ public RevisionStatus getRevisionStatus(String projectId, String revisionId)
-                        .filter(cr -> (RefType.BRANCH == cr.getType()) && isWorkspaceBranchName(cr.getName(), ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE))
-                        .map(cr -> fromWorkspaceBranchName(projectId, cr.getName(), WorkspaceType.USER, ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE))
-                        .collect(Collectors.toList());
+                workspaces = commitRefs.stream().filter(cr -> (RefType.BRANCH == cr.getType()) && isWorkspaceBranchName(cr.getName(), ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE)).map(cr -> fromWorkspaceBranchName(projectId, cr.getName(), WorkspaceType.USER, ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE)).collect(Collectors.toList());
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabRevisionApi.java:
```diff
@@ -202,10 +203,7 @@ public RevisionStatus getRevisionStatus(String projectId, String revisionId)
-                    .filter(cr -> (RefType.TAG == cr.getType()) && isVersionTagName(cr.getName()))
-                    .map(CommitRef::getName)
-                    .collect(Collectors.toList());
+            List<String> versionTagNames = commitRefs.stream().filter(cr -> (RefType.TAG == cr.getType()) && isVersionTagName(cr.getName())).map(CommitRef::getName).collect(Collectors.toList());
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabUserApi.java:
```diff
@@ -55,51 +54,37 @@ public GitLabUserApi(GitLabUserContext userContext)
-                        .map(BaseGitLabApi::fromGitLabAbstractUser)
-                        .forEach(user -> usersById.putIfAbsent(user.getUserId(), user));
-            }
+            PagerTools.stream(getGitLabApi().getUserApi().getUsers(ITEMS_PER_PAGE)).map(BaseGitLabApi::fromGitLabAbstractUser).forEach(user -> usersById.putIfAbsent(user.getUserId(), user));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabUserApi.java:
```diff
@@ -111,57 +96,37 @@ public User getUserById(String userId)
-                        .map(BaseGitLabApi::fromGitLabAbstractUser)
-                        .forEach(user -> usersById.putIfAbsent(user.getUserId(), user));
-            }
+            PagerTools.stream(getGitLabApi().getUserApi().findUsers(search, ITEMS_PER_PAGE)).map(BaseGitLabApi::fromGitLabAbstractUser).forEach(user -> usersById.putIfAbsent(user.getUserId(), user));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -80,11 +48,7 @@ public boolean equals(Object other)
-                patternsEqual(this.artifactIdPattern, that.artifactIdPattern) &&
-                this.disallowedTypes.equals(that.disallowedTypes) &&
-                this.disallowedTypeMessages.equals(that.disallowedTypeMessages);
+        return Objects.equals(this.defaultProjectStructureVersion, that.defaultProjectStructureVersion) && patternsEqual(this.groupIdPattern, that.groupIdPattern) && patternsEqual(this.artifactIdPattern, that.artifactIdPattern);
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -99,36 +63,13 @@ public int hashCode()
-                builder.append(" (\"").append(message).append("\")");
-            }
-        }
+        StringBuilder builder = new StringBuilder("<ProjectCreationConfiguration defaultProjectStructureVersion=").append(this.defaultProjectStructureVersion).append(" groupIdPattern=").append(this.groupIdPattern).append(" artifactIdPattern=").append(this.artifactIdPattern);
```
I think this can be better done with simple string concatenation at this point.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -148,100 +89,31 @@ public Pattern getArtifactIdPattern()
-    public String getDisallowedTypeMessage(ProjectType type)
+    @Deprecated
+    @JsonIgnore
+    public Object getDisallowedTypes()
```
What is the purpose of this method?

akphi replied:
Removed!

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -148,100 +89,31 @@ public Pattern getArtifactIdPattern()
 
     @JsonCreator
-    public static ProjectCreationConfiguration newConfig(@JsonProperty("defaultProjectStructureVersion") Integer defaultProjectStructureVersion, @JsonProperty("groupIdPattern") Pattern groupIdPattern, @JsonProperty("artifactIdPattern") Pattern artifactIdPattern, @JsonProperty("disallowedTypes") Collection<DisallowedType> disallowedTypes)
+    public static ProjectCreationConfiguration newConfig(@JsonProperty("defaultProjectStructureVersion") Integer defaultProjectStructureVersion, @JsonProperty("groupIdPattern") Pattern groupIdPattern, @JsonProperty("artifactIdPattern") Pattern artifactIdPattern)
```
I think we need to keep the "disallowedTypes" property for backward compatibility.

akphi replied:
As discussed, unfortunately, we would not try to keep this backward-compatible

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -148,100 +89,31 @@ public Pattern getArtifactIdPattern()
-        {
-            return new DisallowedType(type, message);
-        }
+        return (pattern1 == pattern2) || ((pattern1 != null) && (pattern2 != null) && (pattern1.flags() == pattern2.flags()) && Objects.equals(pattern1.pattern(), pattern2.pattern()));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/resources/ProjectsResource.java:
```diff
@@ -151,10 +165,8 @@ public void deleteProject(@PathParam("id") String projectId)
-                "importing project " + command.getId() + " (type: " + command.getType() + ", groupId: " + command.getGroupId() + ", artifactId: " + command.getArtifactId() + ")",
-                () -> this.projectApi.importProject(command.getId(), command.getType(), command.getGroupId(), command.getArtifactId())
-        );
+        return executeWithLogging("importing project " + command.getId() + " (type: " + command.getType() + ", groupId: " + command.getGroupId() + ", artifactId: " + command.getArtifactId() + ")",
+            () -> this.projectApi.importProject(command.getId(), command.getGroupId(), command.getArtifactId()));
```
Please revert the formatting to its previous state.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -93,18 +94,15 @@ public GitLabWorkspaceApi(GitLabUserContext userContext, RevisionApi revisionApi
-                    .map(branch -> workspaceBranchToWorkspace(projectId, branch, workspaceType, workspaceAccessType))
-                    .collect(Collectors.toList());
+            Pager<Branch> pager = getGitLabApi().getRepositoryApi().getBranches(gitLabProjectId.getGitLabId(), ITEMS_PER_PAGE);
+            return PagerTools.stream(pager).filter(branch -> (branch != null) && isUserOrGroupWorkspaceBranchName(branch.getName(), workspaceType, workspaceAccessType)).map(branch -> workspaceBranchToWorkspace(projectId, branch, workspaceType, workspaceAccessType)).collect(Collectors.toList());
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -122,18 +120,15 @@ public GitLabWorkspaceApi(GitLabUserContext userContext, RevisionApi revisionApi
-                    .map(branch -> workspaceBranchToWorkspace(projectId, branch, workspaceType, workspaceAccessType))
-                    .collect(Collectors.toList());
+            Pager<Branch> pager = getGitLabApi().getRepositoryApi().getBranches(gitLabProjectId.getGitLabId(), ITEMS_PER_PAGE);
+            return PagerTools.stream(pager).filter(branch -> (branch != null) && isWorkspaceBranchName(branch.getName(), workspaceType, workspaceAccessType)).map(branch -> workspaceBranchToWorkspace(projectId, branch, workspaceType, workspaceAccessType)).collect(Collectors.toList());
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -716,48 +694,28 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(CommitAction.Action.DELETE).withFilePath(diff.getOldPath()));
+                    commitActions.add(new CommitAction().withAction(CommitAction.Action.CREATE).withFilePath(diff.getNewPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -716,48 +694,28 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(CommitAction.Action.CREATE).withFilePath(diff.getNewPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -716,48 +694,28 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getOldPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(CommitAction.Action.UPDATE).withFilePath(diff.getOldPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getOldPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -810,11 +768,7 @@ private boolean attemptToRebaseWorkspaceUsingTemporaryBranch(String projectId, S
-                    mr -> !mr.getRebaseInProgress(),
-                    600,
-                    1000L);
+            CallUntil<MergeRequest, GitLabApiException> rebaseStatusCallUntil = CallUntil.callUntil(() -> withRetries(() -> mergeRequestApi.getRebaseStatus(gitLabProjectId.getGitLabId(), mergeRequest.getIid())), mr -> !mr.getRebaseInProgress(), 600, 1000L);
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -834,22 +788,18 @@ private boolean attemptToRebaseWorkspaceUsingTemporaryBranch(String projectId, S
-                            getWorkspaceBranchName(workspaceId, workspaceType, backupWorkspaceAccessType),
-                            getWorkspaceBranchName(workspaceId, workspaceType, workspaceAccessType),
-                            30, 1_000);
+                    backupBranch = GitLabApiTools.createBranchFromSourceBranchAndVerify(repositoryApi, gitLabProjectId.getGitLabId(), getWorkspaceBranchName(workspaceId, workspaceType, backupWorkspaceAccessType), getWorkspaceBranchName(workspaceId, workspaceType, workspaceAccessType), 30, 1_000);
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -872,28 +822,23 @@ private boolean attemptToRebaseWorkspaceUsingTemporaryBranch(String projectId, S
-                            getWorkspaceBranchName(workspaceId, workspaceType, workspaceAccessType),
-                            tempBranchName,
-                            30, 1_000);
+                    newWorkspaceBranch = GitLabApiTools.createBranchFromSourceBranchAndVerify(repositoryApi, gitLabProjectId.getGitLabId(), getWorkspaceBranchName(workspaceId, workspaceType, workspaceAccessType), tempBranchName, 30, 1_000);
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -1063,53 +1001,33 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(fileToReplace == null ? CommitAction.Action.CREATE : CommitAction.Action.UPDATE).withFilePath(diff.getNewPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -1063,53 +1001,33 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(fileToCreate == null ? CommitAction.Action.CREATE : CommitAction.Action.UPDATE).withFilePath(diff.getNewPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getNewPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -1063,53 +1001,33 @@ else if (diff.getRenamedFile())
-                            .withEncoding(Constants.Encoding.BASE64)
-                            .withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getOldPath()).getContentAsBytes()))
-                    );
+                    commitActions.add(new CommitAction().withAction(fileToUpdate == null ? CommitAction.Action.CREATE : CommitAction.Action.UPDATE).withFilePath(diff.getOldPath()).withEncoding(Constants.Encoding.BASE64).withContent(encodeBase64(workspaceFileAccessContext.getFile(diff.getOldPath()).getContentAsBytes())));
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabWorkspaceApi.java:
```diff
@@ -1063,53 +1001,33 @@ else if (diff.getRenamedFile())
             });
-            commitsApi.createCommit(gitLabProjectId.getGitLabId(), this.getWorkspaceBranchName(workspaceId, workspaceType, conflictResolutionWorkspaceType),
-                    "aggregated changes for conflict resolution", null, null, getCurrentUser(), commitActions);
+            commitsApi.createCommit(gitLabProjectId.getGitLabId(), this.getWorkspaceBranchName(workspaceId, workspaceType, conflictResolutionWorkspaceType), "aggregated changes for conflict resolution", null, null, getCurrentUser(), commitActions);
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabVersionApi.java:
```diff
@@ -362,11 +290,7 @@ private Version newVersion(GitLabProjectId projectId, String revisionId, Version
-                        .filter(GitLabVersionApi::isVersionTagName)
-                        .map(GitLabVersionApi::parseVersionTagName)
-                        .collect(Collectors.toList());
+                List<VersionId> revisionVersionIds = referenceCommitTags.stream().map(CommitRef::getName).filter(GitLabVersionApi::isVersionTagName).map(GitLabVersionApi::parseVersionTagName).collect(Collectors.toList());
```
Please revert to the previous formatting.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/config/LegendSDLCServerFeaturesConfiguration.java:
```diff
@@ -0,0 +1,52 @@
+
+public class LegendSDLCServerFeaturesConfiguration
+{
+    private final Set<ServerSupportedAction> supportedActions;
```
Refactor to `boolean` flag instead

kevin-m-knight-gs reviewed (changes requested):
A number of comments about formatting, but a few substantive comments too.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/config/ProjectCreationConfiguration.java:
```diff
@@ -99,38 +65,14 @@ public int hashCode()
-        builder.append("]>");
-        return builder.toString();
+        return "<ProjectCreationConfiguration defaultProjectStructureVersion=" + this.defaultProjectStructureVersion + " groupIdPattern=" + this.groupIdPattern + " artifactIdPattern=" + this.artifactIdPattern +
+            "]>";
```
I don't think you need the closing square bracket here.
```suggestion
            ">";
```
Also, this could go on the previous line.

kevin-m-knight-gs commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/gitlab/api/GitLabEntityApiTestResource.java:
```diff
@@ -329,23 +326,19 @@ public void runEntitiesInNormalGroupWorkspaceWorkflowTest() throws GitLabApiExce
-                mr -> requiredMergedStatus.equals(mr.getState()),
-                10,
-                500);
+        CallUntil<MergeRequest, GitLabApiException> callUntilMerged = CallUntil.callUntil(() -> mergeRequestApi.getMergeRequest(gitlabProjectId, parsedMergeRequestId), mr -> requiredMergedStatus.equals(mr.getState()), 10, 500);
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java:
```diff
@@ -66,18 +67,10 @@
-            .disable(StreamWriteFeature.AUTO_CLOSE_TARGET)
-            .disable(StreamReadFeature.AUTO_CLOSE_SOURCE)
-            .build();
+    private static final JsonMapper JSON = JsonMapper.builder().enable(SerializationFeature.INDENT_OUTPUT).enable(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS).enable(MapperFeature.SORT_PROPERTIES_ALPHABETICALLY).disable(StreamWriteFeature.AUTO_CLOSE_TARGET).disable(StreamReadFeature.AUTO_CLOSE_SOURCE).build();
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/project/ProjectStructure.java:
```diff
@@ -531,11 +516,7 @@ else if (requireRevisionId)
-                    MetamodelDependency::getMetamodel,
-                    (id, deps) -> (deps.size() > 1) ? deps.stream().collect(StringBuilder::new, (builder, dep) -> dep.appendVersionIdString(builder.append((builder.length() == 0) ? "multiple versions not allowed: " : ", ")), StringBuilder::append).toString() : null,
-                    "metamodels");
+            validateDependencyConflicts(metamodelDependencies, MetamodelDependency::getMetamodel, (id, deps) -> (deps.size() > 1) ? deps.stream().collect(StringBuilder::new, (builder, dep) -> dep.appendVersionIdString(builder.append((builder.length() == 0) ? "multiple versions not allowed: " : ", ")), StringBuilder::append).toString() : null, "metamodels");
```
Please revert to the previous formatting.

kevin-m-knight-gs commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/gitlab/auth/AbstractTestGitLabSession.java:
```diff
@@ -40,91 +37,35 @@ public void testEncoding_Empty()
 
     @Test
-    public void testEncoding_OneOAuthToken()
+    public void testEncoding_OAuthToken()
```
We need at least one OAuth token test with a delay in it.

kevin-m-knight-gs commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/gitlab/auth/AbstractTestGitLabSession.java:
```diff
@@ -40,91 +37,35 @@ public void testEncoding_Empty()
 
     @Test
-    public void testEncoding_OnePrivateAccessToken()
+    public void testEncoding_PrivateAccessToken()
```
We need at least one private access token with a delay in it.

kevin-m-knight-gs commented on legend-sdlc-server/src/test/java/org/finos/legend/sdlc/server/project/TestProjectStructure.java:
```diff
@@ -1475,37 +1317,31 @@ protected void testCantAddInvalidProjectArtifactGenerations(ProjectType projectT
         }
 
-        ProjectConfigurationUpdateBuilder.newBuilder(this.fileAccessProvider, projectType, PROJECT_ID)
+        ProjectConfigurationUpdateBuilder.newBuilder(this.fileAccessProvider, PROJECT_ID).withWorkspace(addGenerationsWorkspaceId, WorkspaceType.USER, ProjectFileAccessProvider.WorkspaceAccessType.WORKSPACE).withMessage("Add a names generations").withArtifactGenerationToAdd(new SimpleArtifactGeneration().withType(ArtifactType.avro).withName("dup")).withProjectStructureExtensionProvider(this.proje ...
```
Please revert to the previous formatting.

akphi commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabProjectApi.java:
```diff
@@ -113,55 +108,22 @@ public Project getProject(String id)
             }
-            // ensure the list of returned projects cannot exceed the limit (if specified) for whatever reasons
-            if (limit != null && projects.size() > limit)
+            if (limit != null && limit > 0)
```
a bit inconsistent with how we're handling other places like `WorkflowApi` or `BuildApi`, I just did a quick search with `limit > 0` throughout the codebase. If you're alright with it, I could make sure the behavior is consistent and throw `400` in other places. Also, will adopt the pattern of usage below:

https://github.com/finos/legend-sdlc/blob/a83bc97e7e578b95b2c3ee83d930d0be3bd13cb8/legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabApiWithFileAccess.java#L724-L730

kevin-m-knight-gs replied:
We should definitely be consistent. I think for 0, we can return an empty result, but for negative values we should throw with a 400 status.

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabApiWithFileAccess.java:
```diff
@@ -721,11 +718,23 @@ public Revision getRevision(String revisionId)
+                }
+                if (limit < 0)
+                {
+                    throw new LegendSDLCServerException("Invalid limit", Status.BAD_REQUEST);
```
```suggestion
                    throw new LegendSDLCServerException("Invalid limit: " + limit, Status.BAD_REQUEST);
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabBuildApi.java:
```diff
@@ -176,16 +177,29 @@ public Build getBuild(String buildId)
+                    }
+                    if (limit < 0)
+                    {
+                        throw new LegendSDLCServerException("Invalid limit", Status.BAD_REQUEST);
```
```suggestion
                        throw new LegendSDLCServerException("Invalid limit: " + limit, Status.BAD_REQUEST);
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitLabProjectApi.java:
```diff
@@ -113,55 +108,36 @@ public Project getProject(String id)
+            if (limit < 0)
             {
-                typesSet = (Set<ProjectType>) types;
+                throw new LegendSDLCServerException("Invalid limit", Status.BAD_REQUEST);
```
```suggestion
                throw new LegendSDLCServerException("Invalid limit: " + limit, Status.BAD_REQUEST);
```

kevin-m-knight-gs commented on legend-sdlc-server/src/main/java/org/finos/legend/sdlc/server/gitlab/api/GitlabWorkflowApi.java:
```diff
@@ -215,9 +218,21 @@ public Workflow getWorkflow(String workflowId)
+                    }
+                    if (limit < 0)
+                    {
+                        throw new LegendSDLCServerException("Invalid limit", Response.Status.BAD_REQUEST);
```
```suggestion
                        throw new LegendSDLCServerException("Invalid limit: " + limit, Response.Status.BAD_REQUEST);
```

kevin-m-knight-gs reviewed (changes requested):
When throwing an invalid limit exception, include the limit the user supplied.
